sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/download/manager/DownloadManagerReleaseDownloader.java

```diff
@@ -0,0 +1,253 @@
+    }
+
+    private synchronized void setDownloadId(long downloadId) {
+        mDownloadId = downloadId;
+        if (mDownloadId != INVALID_DOWNLOAD_IDENTIFIER) {
+            SharedPreferencesManager.putLong(PREFERENCE_KEY_DOWNLOAD_ID, downloadId);
+        } else {
+            SharedPreferencesManager.remove(PREFERENCE_KEY_DOWNLOAD_ID);
+        }
+    }
+
+    @Override
+    public synchronized boolean isDownloading() {
+        return mDownloadId != INVALID_DOWNLOAD_IDENTIFIER;
+    }
+
+    @NonNull
+    @Override
+    public ReleaseDetails getReleaseDetails() {
+        return mReleaseDetails;
+    }
+
+    @Override
+    public void resume() {
+
+        /* If there is active downloads. */
+        long downloadId = getDownloadId();
+        if (downloadId != INVALID_DOWNLOAD_IDENTIFIER) {
+            update();
+        } else {
+            request();
+        }
+    }
+
+    @Override
+    public synchronized void cancel() {
+        if (mRequestTask != null) {
+            mRequestTask.cancel(true);
+            mRequestTask = null;
+        }
+        if (mUpdateTask != null) {
+            mUpdateTask.cancel(true);
+            mUpdateTask = null;
+        }
+        long downloadId = getDownloadId();
+        if (downloadId != INVALID_DOWNLOAD_IDENTIFIER) {
+            remove(downloadId);
+            setDownloadId(INVALID_DOWNLOAD_IDENTIFIER);
+        }
+    }
+
+    /**
+     * Start new download.
+     */
+    private void request() {
+        mRequestTask = AsyncTaskUtils.execute(LOG_TAG, new DownloadManagerRequestTask(this));
+    }
+
+    /**
+     * Update the state on current download.
+     */
+    private void update() {
+        mUpdateTask = AsyncTaskUtils.execute(LOG_TAG, new DownloadManagerUpdateTask(this));
+    }
+
+    private void remove(long downloadId) {
+        AppCenterLog.debug(LOG_TAG, "Removing download and notification id=" + downloadId);
+        AsyncTaskUtils.execute(LOG_TAG, new DownloadManagerRemoveTask(mContext, downloadId));
+    }
+
+    @WorkerThread
+    synchronized void onRequest(DownloadManagerRequestTask task) {
+
+        /* Download file. */
+        Uri downloadUrl = mReleaseDetails.getDownloadUrl();
+        AppCenterLog.debug(LOG_TAG, "Start downloading new release from " + downloadUrl);
+        DownloadManager downloadManager = getDownloadManager();
+        DownloadManager.Request request = new DownloadManager.Request(downloadUrl);
+
+        /* Hide mandatory download to prevent canceling via notification cancel or download UI delete. */
+        if (mReleaseDetails.isMandatoryUpdate()) {
+            request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_HIDDEN);
+            request.setVisibleInDownloadsUi(false);
+        }
+        long enqueueTime = System.currentTimeMillis();
+        long downloadId = downloadManager.enqueue(request);
+
+        /* Check for if state changed and task not canceled in time. */
+        if (mRequestTask == task) {
```

Comment:
All the state checks here don't work anymore.
Before on develop it was using the Distribute lock.
Here it's out of sync with Distribute changing state as this is in a doInBackground.

2 solutions here:

* Since this code on develop was done before the AppCenter looper thing, we were using asynctask for everything but it's not justified anymore. Using AppCenter looper for running code that uses storage is better and avoid state change.
* If AppCenter looper cannot be used, add a method in Distribute that takes a runnable and a release details, synchronize there and execute the runnable if state didn't change.

Either solution could be used for one of my previous comment as well on another place where the refactoring broke synchronization.

Reply:
Added syncs and cancelation checks. Distribute cancels downloader on any state changes